File: src/org/genyris/java/JavaWrapper.java
// Copyright 2010 Peter William Birch <birchb@genyis.org>
//
// This software may be used and distributed according to the terms
// of the Genyris License, in the file "LICENSE", incorporated herein by reference.
//
package org.genyris.java;

import java.lang.reflect.Field;

import org.genyris.core.Atom;
import org.genyris.core.DynamicSymbol;
import org.genyris.core.Exp;
import org.genyris.core.Internable;
import org.genyris.core.Pair;
import org.genyris.core.Symbol;
import org.genyris.core.Visitor;
import org.genyris.exception.GenyrisException;
import org.genyris.interp.Environment;
import org.genyris.interp.UnboundException;

public class JavaWrapper extends Atom {
	private Object _value;

	public Symbol getBuiltinClassSymbol(Internable table) {
		return table.JAVAWRAPPER();
	}

	public JavaWrapper(Object i) {
		_value = i;
	}

	public void acceptVisitor(Visitor guest) throws GenyrisException {
		guest.visitJavaWrapper(this);
	}

	public String toString() {
		return _value.toString();
	}

	public int hashCode() {
		return _value.hashCode();
	}

	public boolean equals(Object compare) {
		if (compare == null)
			return false;
		if (compare.getClass() != this.getClass())
			return false;
		else
			return _value.equals(((JavaWrapper) compare)._value);
	}

	public Exp eval(Environment env) {
		return this;
	}

	public Object getValue() {
		return _value;
	}

	public Environment makeEnvironment(Environment parent)
			throws GenyrisException {
		return new JavaWrapperEnvironment(parent, this);
	}

	public void setField(String fieldName, Exp value, Environment env)
			throws GenyrisException {
		try {
			Field field = _value.getClass().getField(fieldName);
			Object converted = JavaUtils.convertToJava(field.getType(), value, env);
			field.setAccessible(true);
			field.set(_value, converted);
		} catch (SecurityException e) {
			throw new UnboundException(e.getMessage());
		} catch (NoSuchFieldException e) {
			throw new UnboundException(e.getMessage());
		} catch (IllegalArgumentException e) {
			throw new UnboundException(e.getMessage());
		} catch (IllegalAccessException e) {
			throw new UnboundException(e.getMessage());
		}

	}

	public Exp dir(Internable table) {
		Field[] fields = _value.getClass().getFields();
		Exp retval = Pair.cons4(new DynamicSymbol(table.SELF()),
				new DynamicSymbol(table.VARS()),
				new DynamicSymbol(table.CLASSES()), 
				new DynamicSymbol(table.JAVACLASS()), table.NIL());
		for (int i = fields.length - 1; i >= 0; i--) {
			retval = new Pair(new DynamicSymbol(table.internString(fields[i].getName())), retval);
		}
		return retval;
	}

	public Exp getField(Environment env, Symbol sym) throws UnboundException {
		try {
			Field field = _value.getClass().getField(sym.toString());
			field.setAccessible(true);
			return JavaUtils.javaToGenyris(env, field.get(_value));
		} catch (SecurityException e) {
			throw new UnboundException(e.getMessage());
		} catch (NoSuchFieldException e) {
			throw new UnboundException(e.getMessage());
		} catch (IllegalArgumentException e) {
			throw new UnboundException(e.getMessage());
		} catch (IllegalAccessException e) {
			throw new UnboundException(e.getMessage());
		}
	}

	public boolean hasField(Symbol sym) {
		try {
			_value.getClass().getField(sym.toString());
			return true;
		} catch (SecurityException e) {
			return false;
		} catch (NoSuchFieldException e) {
			return false;
		}
	}
    @Override
    public int compareTo(Object o) {
        return this == o ? 0 : 1;
    }

}

File: src/org/genyris/java/JavaWrapper.kt
// Copyright 2010 Peter William Birch <birchb@genyis.org>
//
// This software may be used and distributed according to the terms
// of the Genyris License, in the file "LICENSE", incorporated herein by reference.
//
package org.genyris.java

import org.genyris.core.*
import org.genyris.exception.GenyrisException
import org.genyris.interp.Environment
import org.genyris.interp.UnboundException
import java.lang.reflect.Field

class JavaWrapper(val value: Any) : Atom() {
    override fun getBuiltinClassSymbol(table: Internable): Symbol? {
        return table.JAVAWRAPPER()
    }

    @kotlin.Throws(GenyrisException::class)
    override fun acceptVisitor(guest: Visitor) {
        guest.visitJavaWrapper(this)
    }

    override fun toString(): String {
        return value.toString()
    }

    override fun hashCode(): Int {
        return value.hashCode()
    }

    override fun equals(compare: Any?): Boolean {
        if (compare == null) return false
        if (compare.getClass() != this.getClass()) return false
        else return this.value == (compare as JavaWrapper).value
    }

    override fun compareTo(o: Any): Int {
        if (o.getClass() != this.getClass()) {
            return -1
        }
        if (o is Comparable<*> && o is Comparable<*>) {
            return (this.value as Comparable<*>).compareTo(o)
        }
        return (if (this === o) 0 else 1)
    }

    override fun eval(env: Environment?): Exp {
        return this
    }

    @kotlin.Throws(GenyrisException::class)
    override fun makeEnvironment(parent: Environment?): Environment {
        return JavaWrapperEnvironment(parent, this)
    }

    @kotlin.Throws(GenyrisException::class)
    fun setField(fieldName: String, value: Exp?, env: Environment?) {
        try {
            val field: Field = value.getClass().getField(fieldName)
            val converted = JavaUtils.convertToJava(field.getType(), value, env)
            field.setAccessible(true)
            field.set(this.value, converted)
        } catch (e: SecurityException) {
            throw UnboundException(e.getMessage())
        } catch (e: NoSuchFieldException) {
            throw UnboundException(e.getMessage())
        } catch (e: IllegalArgumentException) {
            throw UnboundException(e.getMessage())
        } catch (e: IllegalAccessException) {
            throw UnboundException(e.getMessage())
        }
    }

    override fun dir(table: Internable): Exp {
        val fields: Array<Field?> = value.getClass().getFields()
        var retval: Exp = Pair.Companion.cons4(
            DynamicSymbol(table.SELF()),
            DynamicSymbol(table.VARS()),
            DynamicSymbol(table.CLASSES()),
            DynamicSymbol(table.JAVACLASS()), table.NIL()
        )
        for (i in fields.indices.reversed()) {
            retval = Pair(DynamicSymbol(table.internString(fields[i]!!.getName())), retval)
        }
        return retval
    }

    @kotlin.Throws(UnboundException::class)
    fun getField(env: Environment, sym: Symbol): Exp? {
        try {
            val field: Field = value.getClass().getField(sym.toString())
            field.setAccessible(true)
            return JavaUtils.javaToGenyris(env, field.get(this.value))
        } catch (e: SecurityException) {
            throw UnboundException(e.getMessage())
        } catch (e: NoSuchFieldException) {
            throw UnboundException(e.getMessage())
        } catch (e: IllegalArgumentException) {
            throw UnboundException(e.getMessage())
        } catch (e: IllegalAccessException) {
            throw UnboundException(e.getMessage())
        }
    }

    fun hasField(sym: Symbol): Boolean {
        try {
            value.getClass().getField(sym.toString())
            return true
        } catch (e: SecurityException) {
            return false
        } catch (e: NoSuchFieldException) {
            return false
        }
    }
}


